### DISCUSSION TITLE
Question about fetching type information

### DISCUSSION THREAD

**Discussion by Andres981231** [microsoft/pyright#3058] [Q&A] [answered]
If I want to write a script which using functions given in pyright-internal to fetch all variables and functions' type in a python file, does the whole process basically be like: tokenize python file->parser to generate parse tree->binder to build scope and symbol table->iterate the parse tree to fetch type information. If so, what functions may I need to use in each step, and I think in the final step, the functions like getType() in typeEvaluator.ts may help. Sorry for the question to be so opening, during my implementing, it seems difficult to get through all complicated functions clearly, so I am here searching for guide from the developer. Thanks a lot.

**Comment by erictraut** [comment, marked as answer]
I recommend starting with [this documentation](https://github.com/microsoft/pyright/blob/main/docs/internals.md). You're going to need to figure out most of this on your own by reading the code and the comments found within it.
